src/Compilers/CSharp/Portable/Lowering/ExtensionMethodReferenceRewriter.cs

```diff
@@ -0,0 +1,149 @@
+            }
+            else
+            {
+                // Rewrite the receiver
+                BoundExpression? rewrittenReceiver = (BoundExpression?)rewriter.Visit(node.ReceiverOpt);
+                rewrittenCall = visitArgumentsAndFinishRewrite(node, rewrittenReceiver);
+            }
+
+            return rewrittenCall;
+
+            BoundExpression visitArgumentsAndFinishRewrite(BoundCall node, BoundExpression? rewrittenReceiver)
+            {
+                return updateCall(
+                    node,
+                    rewriter.VisitMethodSymbol(node.Method),
+                    rewriter.VisitSymbols(node.OriginalMethodsOpt),
+                    rewrittenReceiver,
+                    rewriter.VisitList(node.Arguments),
+                    node.ArgumentRefKindsOpt,
+                    node.InvokedAsExtensionMethod,
+                    rewriter.VisitType(node.Type));
+            }
+
+            static BoundExpression updateCall(
+                BoundCall boundCall,
+                MethodSymbol method,
+                ImmutableArray<MethodSymbol> originalMethodsOpt,
+                BoundExpression? receiverOpt,
+                ImmutableArray<BoundExpression> arguments,
+                ImmutableArray<RefKind> argumentRefKinds,
+                bool invokedAsExtensionMethod,
+                TypeSymbol type)
+            {
+                if (method.OriginalDefinition.ContainingSymbol is NamedTypeSymbol { IsExtension: true } declaringTypeDefinition &&
+                    method.OriginalDefinition.TryGetCorrespondingExtensionImplementationMethod() is MethodSymbol implementationMethod)
+                {
+                    method = implementationMethod.AsMember(method.ContainingSymbol.ContainingType).
+                        ConstructIfGeneric(method.ContainingType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.Concat(method.TypeArgumentsWithAnnotations));
+
+                    if (receiverOpt is not null && arguments.Length == method.ParameterCount - 1)
+                    {
+                        var receiverRefKind = method.Parameters[0].RefKind;
+
+                        if (argumentRefKinds.IsDefault)
+                        {
+                            if (receiverRefKind != RefKind.None)
+                            {
+                                var builder = ArrayBuilder<RefKind>.GetInstance(method.ParameterCount, RefKind.None);
+                                builder[0] = SyntheticBoundNodeFactory.ArgumentRefKindFromParameterRefKind(receiverRefKind, useStrictArgumentRefKinds: false);
+                                argumentRefKinds = builder.ToImmutableAndFree();
+                            }
+                        }
+                        else
+                        {
+                            argumentRefKinds = argumentRefKinds.Insert(0, SyntheticBoundNodeFactory.ArgumentRefKindFromParameterRefKind(receiverRefKind, useStrictArgumentRefKinds: false));
```

Comment:
>false [](http://example.com/codeflow?start=180&length=5)

Consider adding a test for this value

Reply:
> Consider adding a test for this value

Testing for ref receivers is to be done, I added a couple of tests with prototype comment. I am not planning to test more in this PR.

Reply:
BTW, the added tests break for `true`, but quite possibly they take the other branch that has the same logic.

Reply:
Yes, the added tests take the other branch